### ISSUE TITLE
wait container fails to delete sidecar due to race condition

### ISSUE THREAD

**Issue by discordianfish** [argoproj/argo#1390]
**BUG REPORT**

**What happened**:
Sometimes when a workflow finished (observed only for failed workflows but probably unrelated to that), the `wait` container fails to kill the sidecar and shows this in the log:
```
time="2019-06-03T23:06:32Z" level=info msg="Main container completed"
time="2019-06-03T23:06:32Z" level=info msg="Killing sidecars"
time="2019-06-03T23:06:32Z" level=info msg="Annotations monitor stopped"
time="2019-06-03T23:06:32Z" level=info msg="Killing sidecar localstack (796abc1a06044a1e340079b6185f2958a687bc4f507f948b59f1568d5385615d)"
time="2019-06-03T23:06:32Z" level=info msg="Killing sidecar postgres (c470c9385ca3b31333829d3c63073cd11f30965e22726008736ee6c864f395b4)"
time="2019-06-03T23:06:32Z" level=info msg="Killing sidecar redis (086a05c1f2387480c501cd6462f4c17e2948ed39ace86caf83a06465a4bbf0d1)"
time="2019-06-03T23:06:32Z" level=info msg="docker kill --signal TERM 796abc1a06044a1e340079b6185f2958a687bc4f507f948b59f1568d5385615d c470c9385ca3b31333829d3c63073cd11f30965e22726008736ee6c864f395b4 086a05c1f2387480c501cd6462f4c17e2948ed39ace86caf83a06465a4bbf0d1"
time="2019-06-03T23:06:32Z" level=error msg="`docker kill --signal TERM 796abc1a06044a1e340079b6185f2958a687bc4f507f948b59f1568d5385615d c470c9385ca3b31333829d3c63073cd11f30965e22726008736ee6c864f395b4 086a05c1f2387480c501cd6462f4c17e2948ed39ace86caf83a06465a4bbf0d1` failed: Error response from daemon: Cannot kill container: c470c9385ca3b31333829d3c63073cd11f30965e22726008736ee6c864f395b4: Container c470c9385ca3b31333829d3c63073cd11f30965e22726008736ee6c864f395b4 is not running\n"
time="2019-06-03T23:06:32Z" level=error msg="executor error: Error response from daemon: Cannot kill container: c470c9385ca3b31333829d3c63073cd11f30965e22726008736ee6c864f395b4: Container c470c9385ca3b31333829d3c63073cd11f30965e22726008736ee6c864f395b4 is not running\ngithub.com/argoproj/argo/errors.Wrap\n\t/go/src/github.com/argoproj/argo/errors/errors.go:88\ngithub.com/argoproj/argo/errors.InternalWrapError\n\t/go/src/github.com/argoproj/argo/errors/errors.go:71\ngithub.com/argoproj/argo/workflow/executor/docker.(*DockerExecutor).Kill\n\t/go/src/github.com/argoproj/argo/workflow/executor/docker/docker.go:103\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).KillSidecars\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:1080\ngithub.com/argoproj/argo/cmd/argoexec/commands.waitContainer\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:37\ngithub.com/argoproj/argo/cmd/argoexec/commands.NewWaitCommand.func1\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:16\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/src/github.com/spf13/cobra/command.go:766\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/src/github.com/spf13/cobra/command.go:852\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/src/github.com/spf13/cobra/command.go:800\nmain.main\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/main.go:17\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:201\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1333"
time="2019-06-03T23:06:32Z" level=info msg="No output parameters"
```

Apparently the container got restarted:
```
  postgres:
    Container ID:   docker://fdea581572fe0096c768cbe03acffc4cb6268fc7a926a3288f6b3c22ccb5a8ba
    Image:          postgres:9.5
    Image ID:       docker-pullable://postgres@sha256:871f1d9e6d9eccbebfc0744bdb1733170acd9135b5cf0e0f5b68d07ddd02884f
    Port:           <none>
    Host Port:      <none>
    State:          Running
      Started:      Tue, 04 Jun 2019 01:06:32 +0200
    Last State:     Terminated
      Exit Code:    0
      Started:      Mon, 01 Jan 0001 00:00:00 +0000
      Finished:     Mon, 01 Jan 0001 00:00:00 +0000
    Ready:          True
    Restart Count:  1
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from deploy-infra-token-tqqqb (ro)
```

**What you expected to happen**:
The pod should get stopped properly.

**How to reproduce it (as minimally and precisely as possible)**:
Not sure, just happens from time to time, hence I suspect a race condition between the wait container terminating the sidecar and it getting restarted.
**Environment**:
- Argo version: 2.3.0-rc3
- Kubernetes version: 1.11.9

**Comment by abourget** [comment]
I'm seeing the same thing here. But it doesn't seem to me that the container was restarted though.

**Comment by abourget** [comment]
Curiously here, it seems the `wait` container can't watch the `main` container:

```
time="2019-09-27T11:41:37Z" level=info msg="Creating a docker executor"
time="2019-09-27T11:41:37Z" level=info msg="Executor (version: v2.4.0-rc1, build_date: 2019-08-08T22:59:03Z) initialized (pod: default/search-kfdvb-3116563597) with template:\n{\"name\":\"push\",\"arguments\":{},\"inputs\":{},\"outputs\":{},\"metadata\":{},\"container\":{\"name\":\"\",\"image\":\"docker:17.10\",\"command\":[\"sh\",\"-c\"],\"args\":[\"until docker ps\\n  do sleep 3\\ndone\\nsleep 3\\ndocker run --rm debian:latest cat /etc/os-release\\n\"],\"env\":[{\"name\":\"DOCKER_HOST\",\"value\":\"127.0.0.1\"}],\"resources\":{}},\"sidecars\":[{\"name\":\"dind\",\"image\":\"docker:17.10-dind\",\"resources\":{},\"securityContext\":{\"privileged\":true},\"mirrorVolumeMounts\":true}]}"
time="2019-09-27T11:41:37Z" level=info msg="Waiting on main container"
time="2019-09-27T11:41:38Z" level=info msg="main container started with container ID: 20c65710523850ed248c4b06391b9f4aecfc537a90f99c034d535227a5a5b7f1"
time="2019-09-27T11:41:38Z" level=info msg="Starting annotations monitor"
time="2019-09-27T11:41:38Z" level=info msg="docker wait 20c65710523850ed248c4b06391b9f4aecfc537a90f99c034d535227a5a5b7f1"
time="2019-09-27T11:41:38Z" level=info msg="Starting deadline monitor"
time="2019-09-27T11:41:38Z" level=error msg="`docker wait 20c65710523850ed248c4b06391b9f4aecfc537a90f99c034d535227a5a5b7f1` failed: Error response from daemon: No such container: 20c65710523850ed248c4b06391b9f4aecfc537a90f99c034d535227a5a5b7f1\n"
time="2019-09-27T11:41:38Z" level=error msg="executor error: Error response from daemon: No such container: 20c65710523850ed248c4b06391b9f4aecfc537a90f99c034d535227a5a5b7f1\ngithub.com/argoproj/argo/errors.New\n\t/go/src/github.com/argoproj/argo/errors/errors.go:49\ngithub.com/argoproj/argo/errors.InternalError\n\t/go/src/github.com/argoproj/argo/errors/errors.go:60\ngithub.com/argoproj/argo/workflow/common.RunCommand\n\t/go/src/github.com/argoproj/argo/workflow/common/util.go:401\ngithub.com/argoproj/argo/workflow/executor/docker.(*DockerExecutor).Wait\n\t/go/src/github.com/argoproj/argo/workflow/executor/docker/docker.go:95\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).Wait\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:875\ngithub.com/argoproj/argo/cmd/argoexec/commands.waitContainer\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:32\ngithub.com/argoproj/argo/cmd/argoexec/commands.NewWaitCommand.func1\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:16\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/src/github.com/spf13/cobra/command.go:766\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/src/github.com/spf13/cobra/command.go:852\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/src/github.com/spf13/cobra/command.go:800\nmain.main\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/main.go:17\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:201\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1333"
time="2019-09-27T11:41:38Z" level=info msg="Killing sidecars"
time="2019-09-27T11:41:38Z" level=info msg="Annotations monitor stopped"
time="2019-09-27T11:41:38Z" level=info msg="Killing sidecar dind (8d6de8c29aea75a28feb60b799cf413ea6a98d05944b92c3eb0c6914d65aeebd)"
time="2019-09-27T11:41:38Z" level=info msg="docker kill --signal TERM 8d6de8c29aea75a28feb60b799cf413ea6a98d05944b92c3eb0c6914d65aeebd"
time="2019-09-27T11:41:38Z" level=error msg="`docker kill --signal TERM 8d6de8c29aea75a28feb60b799cf413ea6a98d05944b92c3eb0c6914d65aeebd` failed: Error response from daemon: Cannot kill container: 8d6de8c29aea75a28feb60b799cf413ea6a98d05944b92c3eb0c6914d65aeebd: No such container: 8d6de8c29aea75a28feb60b799cf413ea6a98d05944b92c3eb0c6914d65aeebd\n"
time="2019-09-27T11:41:38Z" level=error msg="executor error: Error response from daemon: Cannot kill container: 8d6de8c29aea75a28feb60b799cf413ea6a98d05944b92c3eb0c6914d65aeebd: No such container: 8d6de8c29aea75a28feb60b799cf413ea6a98d05944b92c3eb0c6914d65aeebd\ngithub.com/argoproj/argo/errors.Wrap\n\t/go/src/github.com/argoproj/argo/errors/errors.go:88\ngithub.com/argoproj/argo/errors.InternalWrapError\n\t/go/src/github.com/argoproj/argo/errors/errors.go:71\ngithub.com/argoproj/argo/workflow/executor/docker.(*DockerExecutor).Kill\n\t/go/src/github.com/argoproj/argo/workflow/executor/docker/docker.go:103\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).KillSidecars\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:1084\ngithub.com/argoproj/argo/cmd/argoexec/commands.waitContainer\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:37\ngithub.com/argoproj/argo/cmd/argoexec/commands.NewWaitCommand.func1\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:16\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/src/github.com/spf13/cobra/command.go:766\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/src/github.com/spf13/cobra/command.go:852\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/src/github.com/spf13/cobra/command.go:800\nmain.main\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/main.go:17\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:201\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1333"
time="2019-09-27T11:41:38Z" level=info msg="No output parameters"
time="2019-09-27T11:41:38Z" level=info msg="No output artifacts"
time="2019-09-27T11:41:38Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
time="2019-09-27T11:41:38Z" level=info msg="Alloc=3805 TotalAlloc=11020 Sys=70590 NumGC=4 Goroutines=9"
```

So it's not able to kill the sidecar, but can't monitor the `main` either?

**Comment by abourget** [comment]
Found my issue: I was running the `cos_containerd` images from GKE, switching to `cos` solved the issue.

The problem was that the `executor` in Argo was configured to `docker`, and there's no Docker daemon (or at least, one that runs the images reported by those IDs).. so it was trying to kill images that were run by `containerd`, rather than `docker`.

I also tried the other executors (which are lacking documentation), like `k8sapi`, `kubelet` and `pns`, without success. Each time a different error cropping up.

Hope this helps!
